### PR TITLE
Module 7

### PR DESCRIPTION
**Pull request by luket411**
- [x] You can run all your tests inside Docker following instructions in your README
- [x] You have a passing GitHub Actions workflow that triggers on pushes and pull request runs all your tests (and the pipeline would fail if your tests fail) doesn’t reveal any secret values
- [x] Your submission still satisfies the criteria from previous exercises

----
Additionally Added
- e2e tests (Limited in scope but there to check that they could be run)
- Config to run docker compose with or without corporate proxy settings - `docker-compose.bcn.yml`

### CODE REVIEW

File: Dockerfile [hunk 1]
```diff
@@ -1,4 +1,13 @@
-FROM python:3.12.0a7-buster as base
+ARG BASE_IMAGE=python:3.12.0a7-buster
+ARG START_POINT=${BASE_IMAGE}
+
+# if "base_with_proxy" is passed as build arg, then proxy vars get exported for the image, else they dont.
+FROM ${BASE_IMAGE} as base_with_proxy
+
+ENV http_proxy="http://172.17.0.1:3128"
+ENV https_proxy="http://172.17.0.1:3128"
```
Comment (JackMeadDev) [inline Dockerfile:8]: This looks like a slightly unpleasant workaround to need, but a neat way to achieve it!

File: .github/workflows/style.yml [hunk 2]
```diff
@@ -0,0 +1,26 @@
...
+            - uses: actions/checkout@v2
+
+            - name: Build Docker image
+              run: docker compose build linter
+
+            - name: Make .env stub
+              run: touch .env
+
+            - name: Check todo_app/
+              run: docker compose run linter poetry run pycodestyle --max-line-length=200 todo_app
```
Comment (JackMeadDev) [inline .github/workflows/style.yml]: It feels slightly off that we're using the same docker compose service for both linting as in _checking_ the style, and _fixing_ the style which leads to these slightly ugly commands. I've suggested a possible alternative in the Docker-compose file

File: docker-compose.yml [hunk 3]
```diff
@@ -36,10 +38,46 @@ services:
...
+      - type: bind
+        source: ./todo_app
+        target: /app/todo_app
+    env_file:
+      - ".env"
+    profiles:
+      - e2e_test
+    network_mode: "bridge"
+  
+  linter:
```
Comment (JackMeadDev) [inline docker-compose.yml:74]: As mentioned above, this says it's linting but is actually going a little beyond and formatting the code too - which is great but possibly confusing (and certainly made it non-obvious to me why we were overriding the relevant command in your pipeline!)

One option we could follow is explicitly calling those out, and using the `extends` keyword to minimise the extra compose setup. E.g. we could add:
```yml
  check-style:
    extends: linter
    command: "poetry run pycodestyle --max-line-length=200 todo_app tests e2e_tests"
    profiles:
      - lint
```

(We could similarly have linter extend dev potentially!)

**Review by JackMeadDev — APPROVED** [review]
Nice work, I had no issues (well, apart from Docker misbehaving but that wasn't to do with your code!) running your tests including the lints & e2e tests locally, and I can see everything is passing well in your pipeline 😄 I've left a couple of minor comments, but happy for you to merge when you're ready

Well done in particular going beyond into various stretch exercises too, one thing worth considering as we move towards continuous delivery/deployment will be how you gate on those; the easiest way in GitHub is to manage your jobs within a single workflow file - they can still run in parallel but allows for a simple way to make the deployment dependent on everything passing whereas that might be more fiddly with the multiple-workflow set up!

e.g. might look like:
```yml
jobs:
  job1:
    # details
  job2:
    # details
  job3:
    needs: [job1, job2]
    # details
```

File: Dockerfile (hunk 1 above)
Comment (luket411) [reply Dockerfile:8]: I found this particular setup is needed only for docker compose as docker compose looks in a different place for global proxy settings than docker does. `docker run/build ...` works fine without the work around but `docker compose run/build/up ...` does not.

**Pull request merged**